File: backend/src/models/email.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
#[serde(rename_all = "lowercase")]
pub enum MailProtocol {
    #[default]
    Imap,
    Pop3,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum FilterAction {
    Prioritario,  // siempre notificar, marcar urgente
    Normal,       // clasificar con IA normalmente
    Silencioso,   // clasificar pero no notificar
    Ignorar,      // no procesar, descartarlo
}

impl Default for FilterAction {
    fn default() -> Self {
        Self::Normal
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EmailFilter {
    pub pattern: String,       // ej: "@alumnos.uach.cl", "servicios@uach.cl"
    pub action: FilterAction,
    pub label: String,         // ej: "Estudiantes", "Spam interno"
    #[serde(default)]
    pub auto_tag: Option<String>, // tag para la tarea si se crea
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EmailAccount {
    pub username: String,
    #[serde(alias = "imap_host")]
    pub host: String,
    #[serde(alias = "imap_port")]
    pub port: u16,
    #[serde(default)]
    pub protocol: MailProtocol,
    pub email: String,
    pub password: String,
    #[serde(default)]
    pub filters: Vec<EmailFilter>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EmailMessage {
    pub uid: u32,
    pub from: String,
    pub subject: String,
    pub date: String,
    pub body_preview: String,
    pub ai_classification: Option<String>,
    pub ai_summary: Option<String>,
    pub ai_action: Option<String>,
    #[serde(default)]
    pub filter_label: Option<String>,     // qué filtro le pegó
    #[serde(default)]
    pub filter_action: Option<FilterAction>,
    pub processed: bool,
    pub task_created: bool,
    pub fetched_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct EmailConfig {
    #[serde(default)]
    pub accounts: Vec<EmailAccount>,
    #[serde(default)]
    pub groq_api_key: Option<String>,
}

File: frontend/src/api/email.ts
import { api } from './client'

// --- Email ---

export interface EmailMessage {
  uid: number
  from: string
  subject: string
  date: string
  body_preview: string
  ai_classification: string | null
  ai_summary: string | null
  ai_action: string | null
  filter_label: string | null
  filter_action: string | null
  processed: boolean
  task_created: boolean
  fetched_at: string
}

export interface EmailFilter {
  pattern: string
  action: 'prioritario' | 'normal' | 'silencioso' | 'ignorar'
  label: string
  auto_tag: string | null
}

export async function configureEmailAccount(data: { host: string; port: number; protocol: 'imap' | 'pop3'; email: string; password: string }): Promise<string> {
  const res = await api('/api/email/account', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  })
  if (!res.ok) { const t = await res.text(); throw new Error(t) }
  return res.text()
}

export async function deleteEmailAccount(): Promise<void> {
  const res = await api('/api/email/account', { method: 'DELETE' })
  if (!res.ok) throw new Error('Error al eliminar cuenta')
}

export async function fetchInbox(): Promise<EmailMessage[]> {
  const res = await api('/api/email/inbox')
  if (!res.ok) throw new Error('Error al obtener bandeja')
  return res.json()
}

export async function checkEmailNow(): Promise<string> {
  const res = await api('/api/email/check', { method: 'POST' })
  if (!res.ok) { const t = await res.text(); throw new Error(t) }
  return res.text()
}

export async function classifyEmail(uid: number): Promise<EmailMessage> {
  const res = await api(`/api/email/classify/${uid}`, { method: 'POST' })
  if (!res.ok) throw new Error('Error al clasificar')
  return res.json()
}

export async function emailToTask(uid: number): Promise<string> {
  const res = await api(`/api/email/to-task/${uid}`, { method: 'POST' })
  if (!res.ok) { const t = await res.text(); throw new Error(t) }
  return res.text()
}

export async function setGroqKey(key: string): Promise<string> {
  const res = await api('/api/email/groq-key', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ key }),
  })
  if (!res.ok) throw new Error('Error al configurar key')
  return res.text()
}

export async function fetchEmailFilters(): Promise<EmailFilter[]> {
  const res = await api('/api/email/filters')
  if (!res.ok) throw new Error('Error al obtener filtros')
  return res.json()
}

export async function addEmailFilter(filter: { pattern: string; action: string; label: string; auto_tag?: string }): Promise<void> {
  const res = await api('/api/email/filters', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(filter),
  })
  if (!res.ok) { const t = await res.text(); throw new Error(t) }
}

export async function deleteEmailFilter(pattern: string): Promise<void> {
  const res = await api(`/api/email/filters/${encodeURIComponent(pattern)}`, { method: 'DELETE' })
  if (!res.ok) throw new Error('Error al eliminar filtro')
}
